Update info about Base64 and Unicode (#25253)
mdn/content: files/en-us/glossary/base64/index.md

- Email via [MIME](https://en.wikipedia.org/wiki/MIME)
- Storing complex data in [XML](/en-US/docs/Web/XML).
- Encoding binary data so it can be included in a [`data:` URL](/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs).

## Encoded size increase

Each Base64 digit represents 6 bits of data. So, three 8-bit bytes of the input string/binary file (3×8 bits = 24 bits) can be represented by four 6-bit Base64 digits (4×6 = 24 bits).

This means that the Base64 version of a string or file is typically roughly a third larger than its source (the exact size increase depends on various factors, such as the absolute length of the string, its length modulo 3, and whether padding characters are used).

## JavaScript support

Browsers natively provide two JavaScript functions for decoding and encoding Base64 strings:

- [`btoa`](/en-US/docs/Web/API/btoa): creates a Base64-encoded ASCII string from a string of binary data ("btoa" should be read as "binary to ASCII").
- [`atob`](/en-US/docs/Web/API/atob): decodes a Base64-encoded string ("atob" should be read as "ASCII to binary").

> > **Note:** Base64 is a binary encoding rather than a text encoding, but `btoa` and `atob` were added to the web platform before it supported binary data types. As a result, the two functions use strings to represent binary data, with the codepoint of each character representing the value of each byte. This has led to a common misconception that `btoa` can be used to encode arbitrary text data — for example, creating a Base64 `data:` URL of a text or HTML document.
> >
> > However, the byte-codepoint correspondence only holds true for codepoints up to `0x7f`, and furthermore, codepoints over `0xff` will cause `btoa` to throw an error due to exceeding the maximum value for 1 byte. The next section details ways of working around this limitation when encoding arbitrary Unicode text.

Reviewer: ```suggestion
> **Note:** Base64 is a binary encoding rather than a text encoding, but `btoa` and `atob` were added to the web platform before it supported binary data types. As a result, the two functions use strings to represent binary data, with the code point of each character representing the value of each byte.
>
> This has led to a common misconception that `btoa` can be used to encode arbitrary text data — for example, creating a Base64 `data:` URL of a text or HTML document. However, the byte–character correspondence only holds true for code points up to `0x7f`, and furthermore, code points over `0xff` will cause `btoa` to throw an error due to exceeding the maximum value for 1 byte. The next section details ways of working around this limitation when encoding arbitrary Unicode text.
```

What happens for `0x80` to `0xff`? Is it because the byte is signed?

Author: No, it's not related to being signed — the codepoint of each character corresponds to a raw byte value:

char | byte (hex) | byte (bin)
-|-|-
[NUL control char] | 00 | 0000 0000
A | 41 | 0100 0001
[DEL control char] | 7f | 0111 1111
[control char] | 80 | 1000 0000 (>7 bits, so multibyte in UTF-8)
ÿ | ff | 1111 1111
Ā | 100 | 1 0000 0000 (>8 bits, so `btoa` throws error)

The reason why 8 bits is multibyte in UTF-8 is that UTF-8 is a variable-length encoding, and the first bit being set to 1 signals that it's a multibyte character:

```js
new TextEncoder().encode('\x7f') // Uint8Array [127]
new TextEncoder().encode('\x80') // Uint8Array [194, 128]

new TextDecoder().decode(Uint8Array.from([0x7f])) // '\x7F' - OK
new TextDecoder().decode(Uint8Array.from([0x80])) // '�' - replacement character, because an 0x80 byte on its own isn't valid UTF-8
```

I'll change the wording to "byte-to-UTF-8 correspondence" as that's more accurate. Maybe it'd be better to add more explanation too, but I'm not sure how to do so in a way that's concise, correct, and not too tangential to the main point of the article.

## The "Unicode Problem"

Since `btoa` only accepts characters within the `Latin1` Unicode range (code points `0x00` to `0xff`), calling `btoa` on a string will cause a "Character Out Of Range" exception if a character exceeds that range. For use cases where you need to encode arbitrary Unicode text, it is necessary to first convert the string to its constituent bytes in UTF-8, and then encode the bytes.

### Solution 1 – `TextEncoder` and `TextDecoder`

The simplest solution is to use `TextEncoder` and `TextDecoder` to convert between UTF-8 and single-byte representations of the string:

```js
function base64ToBytes(base64) {
  const binString = atob(base64);
  return Uint8Array.from(binString, (m) => m.codePointAt(0));
}

function bytesToBase64(bytes) {
  const binString = Array.from(bytes, (x) => String.fromCodePoint(x)).join("");
  return btoa(binString);
}

// Usage
bytesToBase64(new TextEncoder().encode("a Ā 𐀀 文 🦄")); // "YSDEgCDwkICAIOaWhyDwn6aE"
new TextDecoder().decode(base64ToBytes("YSDEgCDwkICAIOaWhyDwn6aE")); // "a Ā 𐀀 文 🦄"
```

> ### Solution 2 – custom Base64 implementation

Reviewer: Let's get rid of this (sorry I know you probably put a lot of effort into refactoring!) All reasonable browser versions support the features above, and this chunk is too unmaintainable and error-prone.

Author: No worries, will remove — I mainly just renamed some things and removed the hungarian notation on the variable names.

It's true front-end near-universally supports the features. On back-end, both Node and Deno support `TextEncoder`/`TextDecoder`, but only Deno supports `atob`/`btoa`. In any case, Node typically seems to use `Buffer` to do Base64 conversion, rather than any special libraries or polyfills.

**Base64** is a group of similar [binary-to-text encoding](https://en.wikipedia.org/wiki/Binary-to-text_encoding) schemes that represent binary data in an ASCII string format by translating it into a radix-64 representation. The term _Base64_ originates from a specific [MIME content transfer encoding](https://en.wikipedia.org/wiki/MIME#Content-Transfer-Encoding).

When the term "Base64" is used on its own to refer to a specific algorithm, it typically refers to the version of Base64 outlined in [RFC 4648](https://datatracker.ietf.org/doc/html/rfc4648), section 4, which uses the following alphabet to represent the radix-64 digits, alongside `=` as a padding character:

> ```plain

Reviewer: ~~~suggestion
```
~~~

ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/
```

A common variant is "Base64 URL safe", which omits the padding and replaces `+/` with `-_` to avoid characters that might cause problems in URL path segments or query parameters.

Base64 encoding schemes are commonly used to encode binary data for storage or transfer over media that can only deal with ASCII text (or some superset of ASCII that still falls short of accepting arbitrary binary data). This ensures that the data remains intact without modification during transport. Common applications of Base64 include:

- Email via [MIME](https://en.wikipedia.org/wiki/MIME)

> - Storing complex data in [XML](/en-US/docs/Web/XML).
> - Encoding binary data so it can be included in a [`data:` URL](/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs).

Reviewer: ```suggestion
- Storing complex data in [XML](/en-US/docs/Web/XML)
- Encoding binary data so it can be included in a [`data:` URL](/en-US/docs/Web/HTTP/Basics_of_HTTP/Data_URLs)
```

This means that the Base64 version of a string or file is typically roughly a third larger than its source (the exact size increase depends on various factors, such as the absolute length of the string, its length modulo 3, and whether padding characters are used).

## JavaScript support

Browsers natively provide two JavaScript functions for decoding and encoding Base64 strings:

- [`btoa`](/en-US/docs/Web/API/btoa): creates a Base64-encoded ASCII string from a string of binary data ("btoa" should be read as "binary to ASCII").
- [`atob`](/en-US/docs/Web/API/atob): decodes a Base64-encoded string ("atob" should be read as "ASCII to binary").

> **Note:** Base64 is a binary encoding rather than a text encoding, but `btoa` and `atob` were added to the web platform before it supported binary data types. As a result, the two functions use strings to represent binary data, with the code point of each character representing the value of each byte. This has led to a common misconception that `btoa` can be used to encode arbitrary text data — for example, creating a Base64 `data:` URL of a text or HTML document.
>

> > However, in [UTF-8](/en-US/docs/Glossary/UTF-8), the byte-to-code-point correspondence only holds true for code points up to `0x7f`. Furthermore, code points over `0xff` will cause `btoa` to throw an error due to exceeding the maximum value for 1 byte. The next section details ways of working around this limitation when encoding arbitrary Unicode text.

Reviewer: TBH, I don't think UTF-8 becomes a problem with `btoa` because JavaScript strings are conceptually UTF-16, so the only requirement for the input is that the most significant 8 bits of each UTF-16 code unit are all 0, right?

Author: JS strings are UTF-16 internally and always will be for backward-compatibility reasons, but UTF-8 is the most common encoding for web content and is [recommended by the W3C for interoperability](https://www.w3.org/International/questions/qa-choosing-encodings#:~:text=Choose%20UTF%2D8%20for%20all%20content%20and%20consider%20converting%20any%20content%20in%20legacy%20encodings%20to%20UTF%2D8). UTF-8 is also the only encoding supported by `TextEncoder/TextDecoder` for the same reason.

> the only requirement for the input is that the most significant 8 bits of each UTF-16 code unit are all 0, right?

That's one way of conceptualizing it, and quite likely is how it's typically implemented in browser engines. The binary representation of any code point `0..0xff` is identical to its most significant 8 bits in UTF-16.

I think I'll just re-word this bit again as the UTF-8 reference here makes it more confusing, not less.

The simplest solution is to use `TextEncoder` and `TextDecoder` to convert between UTF-8 and single-byte representations of the string:

```js
function base64ToBytes(base64) {
  const binString = atob(base64);
  return Uint8Array.from(binString, (m) => m.codePointAt(0));
}

function bytesToBase64(bytes) {
  const binString = Array.from(bytes, (x) => String.fromCodePoint(x)).join("");
  return btoa(binString);
}

// Usage
bytesToBase64(new TextEncoder().encode("a Ā 𐀀 文 🦄")); // "YSDEgCDwkICAIOaWhyDwn6aE"
new TextDecoder().decode(base64ToBytes("YSDEgCDwkICAIOaWhyDwn6aE")); // "a Ā 𐀀 文 🦄"
```

## Converting arbitrary binary data

The `bytesToBase64` and `base64ToBytes` functions in the previous section can be used directly to convert between Base64 strings and [`Uint8Array`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array)s.

Alternatively, asynchronous conversion between base64 data URLs is possible natively within the web platform via the [`FileReader`](/en-US/docs/Web/API/FileReader) and [`fetch`](/en-US/docs/Web/API/Fetch_API) APIs:

```js

> async function bytesToBase64DataUrl(bytes, type = "application/octet-stream") {

Reviewer: Could you make "URL" all-caps everywhere?

Author: Can do, but it doesn't seem like there's a consistent convention for this on MDN? The [MDN JS style guide](https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Writing_style_guide/Code_style_guide/JavaScript) only specifies "camel case", but none of the examples contain abbreviations. Case-sensitive regex search for `[a-z]URL` in all `.md` files gives 873 results, whereas `[a-z]Url` only gives 288; however, a lot of the `[a-z]URL` results are for web-platform APIs, not userland functions. And the web platform also contains `XMLHttpRequest` 😉

FWIW, I think Google's "[(nearly) deterministic scheme](https://google.github.io/styleguide/jsguide.html#naming-camel-case-defined)" for camel-casing words is a pretty good convention.

Reviewer: There is `readAsDataURL` on this page, so I just wanted it to be consistent. Outside of JS docs, I don't have opinions that are as strong, so it's okay if you prefer otherwise.

Author: `readAsDataURL` is part of a web-platform API (method of `FileReader`), so no way of changing that. The web platform continues to use force-capitalization even with new APIs, presumably for historical consistency, but userland code/libraries often break with the convention.

A couple of arguments against force-capitalizing abbreviations in camel case in userland code — consistently doing so prevents using using capitalization as a semantic differentiator between consts, classes, and instances. I.e. with true camel case, you can do something like `svg = new Svg()`, whereas with force-capitalized camel case, you'd have to do something like `SVGInstance = new SVG()`. Also, you end up with things like `let HTTPSURL = "https://example.com"` (HTTPS URL? Or HTTP SURL? Also, isn't that supposed to be a constant?) versus `let httpsUrl = "https://example.com"`

Reviewer: Sure, I'm okay with this. FWIW, all-caps camelCase style will usually write `svgInstance` (i.e. the capitalization should be _consistent_)

In JavaScript there are two functions respectively for decoding and encoding Base64 strings:

- [`btoa()`](/en-US/docs/Web/API/btoa): creates a Base64-encoded ASCII string from a string of binary data ("btoa" should be read as "binary to ASCII").
- [`atob()`](/en-US/docs/Web/API/atob): decodes a Base64-encoded string ("atob" should be read as "ASCII to binary").

The algorithm used by `atob()` and `btoa()` is specified in [RFC 4648](https://datatracker.ietf.org/doc/html/rfc4648), section 4.

> > **Note:** `btoa()` expects to be passed a string of "binary" data and will throw an exception if the given string contains any characters whose codepoint exceeds `0xff` (the maximum value that can be held in 1 byte). See [§ The "Unicode Problem"](#the_unicode_problem) for ways of working around this limitation.

Reviewer: ```suggestion
> **Note:** `btoa()` expects a string of "binary" data and will throw an exception if the given string contains any character whose code point exceeds `0xff` (the maximum value that can be held in 1 byte). See [The "Unicode Problem"](#the_unicode_problem) for ways of working around this limitation.
```

## The "Unicode Problem"

Since `window.btoa` only accepts characters within the `Latin1` Unicode range (codepoints `0x00..0xff`), calling `window.btoa` on a string will cause a `Character Out Of Range` exception if a character exceeds that range. There are two possible methods to solve this problem:

- Convert the string to its constituent bytes in a Unicode encoding, such as UTF-8, then encode the bytes;
- Escape the whole string and then encode it.

> ### Converting the string to UTF-8

Reviewer: I don't see a second H3 heading in this page. We don't create single subheadings.

Author: Thanks for catching! Fixed now — § "Escaping the string before encoding it" should also be H3

#### Solution 1 – `TextEncoder` and `TextDecoder`

```js

> const encodeB64 = (str) => {

Reviewer: ```suggestion
function encodeB64(str) {
```

  const utf8Bytes = new TextEncoder().encode(str);
  const asciiCodePoints = String.fromCodePoint(...utf8Bytes);

  return btoa(asciiCodePoints);
}

// usage

> encodeB64('á ü 文 🦄') // 'w6Egw7wg5paHIPCfpoQ='

Reviewer: ```suggestion
encodeB64("á ü 文 🦄"); // 'w6Egw7wg5paHIPCfpoQ='
```

```

```js

> const decodeB64 = (b64) => {

Reviewer: ```suggestion
function decodeB64(b64) {
```

  const asciiCodePoints = atob(b64);
  const utf8Bytes = new Uint8Array(asciiCodePoints.split('').map((char) => char.codePointAt(0)));

  return new TextDecoder().decode(utf8Bytes);
}

// usage

> encodeB64('w6Egw7wg5paHIPCfpoQ=') // 'á ü 文 🦄'

Reviewer: ```suggestion
encodeB64("w6Egw7wg5paHIPCfpoQ="); // 'á ü 文 🦄'
```

## The "Unicode Problem"

> Since `btoa` only accepts characters within the `Latin1` Unicode range (codepoints `0x00..0xff`), calling `btoa` on a string will cause a `Character Out Of Range` exception if a character exceeds that range. There are two possible methods to solve this problem:

Reviewer: ```suggestion
Since `btoa` only accepts characters within the `Latin1` Unicode range (code points `0x00` to `0xff`), calling `btoa` on a string will cause a "Character Out Of Range" exception if a character exceeds that range. There are two possible methods to solve this problem:
```

> - Convert the string to its constituent bytes in a Unicode encoding, such as UTF-8, then encode the bytes;

Reviewer: ```suggestion
- Convert the string to its constituent bytes in a Unicode encoding, such as UTF-8, and then encode the bytes;
```

- Escape the whole string and then encode it.

### Converting the string to UTF-8

#### Solution 1 – `TextEncoder` and `TextDecoder`

```js
function encodeBase64(unicodeString) {
  const utf8Bytes = new TextEncoder().encode(unicodeString);

>   const binString = Array.from(utf8Bytes, (x) => String.fromCodePoint(x)).join(
>     ""
>   );

Reviewer: ```suggestion
  const binString = Array.from(utf8Bytes, String.fromCodePoint).join("");
```

Author: Not equivalent — `String.fromCodePoint` can take multiple arguments (it should really have been called `String.fromCodePoints`), and `Array#map` supplies it with a second number (the index).

So whereas `Array.from([0x41, 0x42, 0x43], (x) => String.fromCodePoint(x)).join("")` gives `"ABC"`, `Array.from([0x41, 0x42, 0x43], String.fromCodePoint).join("")` instead gives `"A\x00B\x01C\x02"`.

alert(sMyOutput);
```

### Escaping the string before encoding it

```js
function utf8_to_b64(str) {
  return window.btoa(unescape(encodeURIComponent(str)));
}

function b64_to_utf8(str) {
  return decodeURIComponent(escape(window.atob(str)));
}

// Usage:
utf8_to_b64("✓ à la mode"); // "4pyTIMOgIGxhIG1vZGU="
b64_to_utf8("4pyTIMOgIGxhIG1vZGU="); // "✓ à la mode"
```

This solution has been proposed by [Johan Sundström](https://ecmanaut.blogspot.com/2006/07/encoding-decoding-utf8-in-javascript.html).

Another possible solution without utilizing the now deprecated 'unescape' and 'escape' functions.
This alternative, though, does not perform base64 encoding of the input string.
Note the differences in the outputs of `utf8_to_b64` and `b64EncodeUnicode`.
Adopting this alternative may lead to interoperability issues with other applications.

```js
function b64EncodeUnicode(str) {
  return btoa(encodeURIComponent(str));
}

function UnicodeDecodeB64(str) {
  return decodeURIComponent(atob(str));
}

b64EncodeUnicode("✓ à la mode"); // "JUUyJTlDJTkzJTIwJUMzJUEwJTIwbGElMjBtb2Rl"
UnicodeDecodeB64("JUUyJTlDJTkzJTIwJUMzJUEwJTIwbGElMjBtb2Rl"); // "✓ à la mode"

> ```

Reviewer: I don't think we should mention this in a "base64" glossary because as you found it's not base64. There's no risk in using `unescape` and we should just say "it's acceptable in this case compared to writing your own implementation of the same thing"

Author: @Josh-Cena I've gone ahead and removed all the code except my `TextEncoder` and `TextDecoder` solutions and the custom UTF-8 one.

Arguably `TextEncoder` and `TextDecoder` have wide enough support to make even the custom UTF-8 one unnecessary, but I think it's at least got some pedagogical value. The custom decode is also significantly more performant in Chrome, and the custom encode also seems slightly more performant, although there's not much in it.

### Appendix: Decode a Base64 string to Uint8Array or ArrayBuffer

> The functions in [§ Solution 2 – custom UTF-8 implementation using `TypedArray`s](#solution_2_–_custom_utf-8_implementation_using_typedarrays) also let us create [uint8Arrays](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array) or [arrayBuffers](/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer) from Base64-encoded strings:

Reviewer: ```suggestion
The functions in [Solution 2 – custom UTF-8 implementation using `TypedArray`s](#solution_2_–_custom_utf-8_implementation_using_typedarrays) also let us create [`Uint8Array`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array) or [`ArrayBuffer`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/ArrayBuffer) objects from Base64-encoded strings:
```

```

> **Note:** The function `base64DecToArr(sBase64[, nBlocksSize])` returns

> > a [`uint8Array`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array) of bytes.

Reviewer: ```suggestion
> a [`Uint8Array`](/en-US/docs/Web/JavaScript/Reference/Global_Objects/Uint8Array) of bytes.
```
